external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/CachedKafkaProducer.scala

```diff
@@ -40,81 +90,114 @@ private[kafka010] object CachedKafkaProducer extends Logging {
-    kafkaProducer
-  }
+      .build[Seq[(String, Object)], CachedKafkaProducer](cacheLoader)
 
   /**
    * Get a cached KafkaProducer for a given configuration. If matching KafkaProducer doesn't
    * exist, a new KafkaProducer will be created. KafkaProducer is thread safe, it is best to keep
    * one instance per specified kafkaParams.
    */
-  private[kafka010] def getOrCreate(kafkaParams: ju.Map[String, Object]): Producer = {
-    val paramsSeq: Seq[(String, Object)] = paramsToSeq(kafkaParams)
+  private[kafka010] def acquire(kafkaParamsMap: ju.Map[String, Object]): CachedKafkaProducer = {
+    val paramsSeq: Seq[(String, Object)] = paramsToSeq(kafkaParamsMap)
     try {
-      guavaCache.get(paramsSeq)
+      val producer = this.synchronized {
+        val cachedKafkaProducer: CachedKafkaProducer = guavaCache.get(paramsSeq)
+        cachedKafkaProducer.inUseCount.incrementAndGet()
+        logDebug(s"Granted producer $cachedKafkaProducer")
+        cachedKafkaProducer
+      }
+      producer
     } catch {
-      case e @ (_: ExecutionException | _: UncheckedExecutionException | _: ExecutionError)
+      case e@(_: ExecutionException | _: UncheckedExecutionException | _: ExecutionError)
         if e.getCause != null =>
         throw e.getCause
     }
   }
 
-  private def paramsToSeq(kafkaParams: ju.Map[String, Object]): Seq[(String, Object)] = {
-    val paramsSeq: Seq[(String, Object)] = kafkaParams.asScala.toSeq.sortBy(x => x._1)
+  private def paramsToSeq(kafkaParamsMap: ju.Map[String, Object]): Seq[(String, Object)] = {
+    val paramsSeq: Seq[(String, Object)] = kafkaParamsMap.asScala.toSeq.sortBy(x => x._1)
     paramsSeq
   }
 
-  /** For explicitly closing kafka producer */
-  private[kafka010] def close(kafkaParams: ju.Map[String, Object]): Unit = {
-    val paramsSeq = paramsToSeq(kafkaParams)
-    guavaCache.invalidate(paramsSeq)
+  /* Release a kafka producer back to the kafka cache. We simply decrement it's inuse count. */
+  private[kafka010] def release(producer: CachedKafkaProducer, failing: Boolean): Unit = {
+    this.synchronized {
+      // It should be ok to call release multiple times on the same producer object.
+      if (producer.inUse()) {
+        // So that we do not end up with -ve in-use counts.
+        producer.inUseCount.decrementAndGet()
+        logDebug(s"Released producer $producer.")
+      } else {
+        logWarning(s"Tried to release a not in use producer, $producer.")
+      }
+      if (failing) {
+        // If this producer is failing to write, we remove it from cache.
+        // So that it is re-created, eventually.
+        val cachedProducer = guavaCache.getIfPresent(producer.kafkaParams)
+        if (cachedProducer != null && cachedProducer.id == producer.id) {
+          logDebug(s"Invalidating a failing producer: $producer.")
+          guavaCache.invalidate(producer.kafkaParams)
+        }
+      }
+    }
+    if (!producer.inUse() && !producer.isCached) {
```

Comment:
And I feel complicated that there're two different places which close producer. 

Every producers are cached, and all the producers will be go through removal listener to be removed in cache and be closed immediately or added to close queue to be closed eventually (when it fulfills the condition).

Then from outer removal listener we just need to concern about how to handle close queue. The if statement here makes us to determine when to handle close queue immediately, but there's race condition I described before so it's not exactly correct. 
(Actually if it's exactly thread-safe we even don't need a close queue. Just release it once if conditions are met.)

Other approach would be injecting cleanup of close queue to release with some probability (Guava cache seems to evict like this way), or even periodically in background thread. Handling close queue doesn't need to be synchronized with others (if we avoid concurrent modification exception) - only thing would be `inUse` in producer, but once we revisit the producer eventually (not too late) no need to be exact.

Reply:
I realized `close` which removal listener calls is same as `close` here so the possibility is pretty higher than I thought. (I feel it might be too frequent though.)

Then I think if statement here can be just removed and it wouldn't hurt.

Reply:
The race condition is not clear to me, you have concluded in the comment that there is no problem, since eventually producer will be closed. 
either a producer is closed immediately, because it is not in use and evicted by guava, or it is in use and should be closed once it is finally released by all the threads using it.

Can you give a test, to reproduce the problem you are stating?

Reply:
Could you point out which you don't understand my explanation of race condition?

In short, `Adding producer to closeQueue` and `marking producer as uncached` are not being done atomically given there's no synchronization in `removalListener.onRemoval`. There's a case producer is added to closeQueue but isCached is still `true` - race condition happens here.

In other words, suppose `inCount` in producer is exactly correct and thread-safe: then why we need close queue? There should be "exactly only once" when inUse = false, either `onRemoval` or `release`, then closeQueue doesn't be needed and just closing the producer should work. `closeQueue` is just for mitigation when inCount is off, and looping closeQueue every time when other producer is closing is not free.

Reply:
The purpose of closeQueue is, when an inuse producer is evicted by cace, due to a long running task. Then in such rare cases, closeQueue will hold the producer until it is finally released by all the task. Two important properties to note, 1) while the producer exists in the closeQueue, it cannot be acquired by any new tasks. 2) Once all the tasks have released it, then it closed and removed from the closeQueue. 

Currently, the answers to the following question is not clear. Please feel free to help me, anyone.

How closeQueue is just for mitigation of inCount goes out of sync?

How does race condition happens when isCached is true and object is not added to closeQueue Or vice-versa?

Reply:
@HeartSaVioR nice catch again. More eyes more opportunities.

@ScrapCodes To oversimplify this problem: because of threading producers sitting in `closeQueue` may wait really long to be closed (until another `release` called on another producer no race condition happens again). Since its race condition don't think meaningful test can be given and even test is written maybe the code hits it rarely. On the other hand such race in streaming is nasty enough to spend weeks to hunt this down.

I personally like the way how guava is doing cleanup. Simple and effective.

Reply:
@ScrapCodes I can't really give better explanation than @HeartSaVioR given [here](https://github.com/apache/spark/pull/19096#discussion_r270608975). `removalListener` runs in separate thread than `release` for instance.

Reply:
@ScrapCodes 

> Then in such rare cases, closeQueue will hold the producer until it is finally released by all the tasks.

If things pretty work fine, closeQueue is not needed: each task will hold the producer when it uses the producer even closeQueue doesn't exist, and eventually one of task will close it when decreasing inCount makes inCount being zero.

> 1) As soon as the producer enters in the closeQueue, it cannot be acquired by any new tasks.(Guarantee by guava) 

I think it should be guaranteed by Guava when onRemoval is called with such producer. No one can get the producer afterwards. Same without closeQueue, right?

> 2) Once all the tasks have released it, then it is closed and removed from the closeQueue.

closeQueue is not needed as I explained earlier.

> How closeQueue is just for mitigation of inCount goes out of sync?

As I explained race condition can happen, let's assume inCount can be out of sync. Once producer is being added to closeQueue, even race condition I've specified happens, the producer is added to closeQueue (though it will not be closed at that time) and it can be closed once another producer is being closed. So getting another chance to be closed: I meant this as "mitigation".

> How does race condition happens when isCached is true and object is not added to closeQueue Or vice-versa?

I've already explained two times and @gaborgsomogyi can also see the case. Please take a deep look at my comment: I understand it's not easy for author to see defect in his/her own code.

Reply:
Maybe just removing closeQueue might resolve the race condition. (Though I guess we still need to synchronize between `release` and `removal listener` to ensure safety) The issue happens because there're two steps to uncache the producer.

Reply:
> Maybe just removing closeQueue might resolve the race condition.

@HeartSaVioR At the first glance don't really see how to solve it without `closeQueue`. If the task failed failed for instance (maybe producer issue maybe not) several other users not yet finished so cannot be closed immediately. Yeah, there is a reference to this consumer on the user side but since there is no guarantee they will call release with that specific producer I think it's better to have this queue.

I see the solution in the proper synchronization. As an optional change I like how guava does the clean-up and I would move this code:
```
    for (p <- closeQueue.iterator().asScala) {
      if (!p.inUse()) {
        closeQueue.remove(p)
        p.close()
      }
    }
```
at the end of `release` without any condition. If its empty doesn't consume much ticks and more safe to deal with producers in the queue.

Reply:
@gaborgsomogyi 

> Yeah, there is a reference to this consumer on the user side but since there is no guarantee they will call release with that specific producer I think it's better to have this queue.

The situation is same even having closeQueue. If we suppose the case any borrower doesn't call release (due to failure - say, leak on releasing resource), producer never reaches inCount = 0 and producer wouldn't be closed in closeQueue.

Reply:
I see the point now and makes sense. This case both way is fine for me.

Reply:
Maybe close queue is helpful, as we can delegate a close to be done outside the synchronized block.